### PR TITLE
Fixed #27863 -- Added support for the SameSite cookie flag.

### PR DESCRIPTION
**Pull request by kravietz**
Resolving https://code.djangoproject.com/ticket/27863

### CODE REVIEW

File: django/conf/global_settings.py [hunk 1]
```diff
@@ -157,6 +157,7 @@ def gettext_noop(s):
 LANGUAGE_COOKIE_AGE = None
 LANGUAGE_COOKIE_DOMAIN = None
 LANGUAGE_COOKIE_PATH = '/'
+LANGUAGE_COOKIE_SAMESITE = 'lax'
```
Comment (charettes) [inline django/conf/global_settings.py]: I'm not convinced this is worth adding for the language cookie as it's not carrying any critical data (notice how `LANGUAGE_COOKIE_SECURE` and `LANGUAGE_COOKIE_HTTPONLY` don't exist either.

File: django/contrib/messages/storage/cookie.py [hunk 2]
```diff
@@ -86,6 +86,7 @@ def _update_cookie(self, encoded_data, response):
                 domain=settings.SESSION_COOKIE_DOMAIN,
                 secure=settings.SESSION_COOKIE_SECURE or None,
                 httponly=settings.SESSION_COOKIE_HTTPONLY or None,
+                samesite=settings.SESSION_COOKIE_SAMESITE or None
```
Comment (charettes) [inline django/contrib/messages/storage/cookie.py]: Add a trailing comma.

File: django/contrib/sessions/middleware.py [hunk 3]
```diff
@@ -69,5 +69,6 @@ def process_response(self, request, response):
                             path=settings.SESSION_COOKIE_PATH,
                             secure=settings.SESSION_COOKIE_SECURE or None,
                             httponly=settings.SESSION_COOKIE_HTTPONLY or None,
+                            samesite=settings.SESSION_COOKIE_SAMESITE or None
```
Comment (charettes) [inline django/contrib/sessions/middleware.py]: Add a trailing comma.

File: django/http/response.py [hunk 4]
```diff
@@ -198,6 +198,8 @@ def set_cookie(self, key, value='', max_age=None, expires=None, path='/',
             self.cookies[key]['secure'] = True
         if httponly:
             self.cookies[key]['httponly'] = True
+        if samesite:
+            self.cookies[key]['samesite'] = samesite.capitalize()
```
Comment (charettes) [inline django/http/response.py]: what's the reason for capitalizing it?

File: django/middleware/csrf.py [hunk 5]
```diff
@@ -190,6 +190,7 @@ def _set_token(self, request, response):
                 path=settings.CSRF_COOKIE_PATH,
                 secure=settings.CSRF_COOKIE_SECURE,
                 httponly=settings.CSRF_COOKIE_HTTPONLY,
+                samesite=settings.CSRF_COOKIE_SAMESITE
```
Comment (charettes) [inline django/middleware/csrf.py]: Add a trailing comma.

File: django/views/i18n.py [hunk 6]
```diff
@@ -55,6 +55,7 @@ def set_language(request):
                     max_age=settings.LANGUAGE_COOKIE_AGE,
                     path=settings.LANGUAGE_COOKIE_PATH,
                     domain=settings.LANGUAGE_COOKIE_DOMAIN,
+                    samesite=settings.LANGUAGE_COOKIE_SAMESITE
```
Comment (charettes) [inline django/views/i18n.py]: You can remove references to `LANGUAGE_COOKIE_SAMESITE`.

File: docs/ref/settings.txt [hunk 7]
```diff
@@ -1701,6 +1713,18 @@ permanently (via the :setting:`LANGUAGE_COOKIE_NAME` setting) and to add
...
+``LANGUAGE_COOKIE_SAMESITE``
+--------------------------
+
+Default: ``'lax'``
+
+The value of ``SameSite`` flag on the languasge cookie. This flag prevents the cookie
+from being sent in cross-site requests.
+
+See :setting:`SESSION_COOKIE_SAMESITE` for details on ``SameSite``.
+
```
Comment (charettes) [inline docs/ref/settings.txt]: You can remove reference to `LANGUAGE_COOKIE_SAMESITE`.

File: tests/sessions_tests/tests.py [hunk 8]
```diff
@@ -659,6 +659,21 @@ def test_httponly_session_cookie(self):
...
+        middleware = SessionMiddleware()
+
+        # Simulate a request the modifies the session
+        middleware.process_request(request)
+        request.session['hello'] = 'world'
+
+        # Handle the response through the middleware
+        response = middleware.process_response(request, response)
+        self.assertTrue(
+            response.cookies[settings.SESSION_COOKIE_NAME]['samesite'])
```
Comment (charettes) [inline tests/sessions_tests/tests.py]: Test the explicit value. This should fit on a single line.

File: django/conf/global_settings.py (hunk 1 above)
Comment (alex) [inline django/conf/global_settings.py]: I wouldn't bother, there's nothing bad you can do with cross-origin language headers.

**Comment by alex** [comment]
Thanks for working on this!

File: django/http/response.py (hunk 4 above)
Comment (kravietz) [inline django/http/response.py]: Preserve the capitalization from the original standard.

Comment (charettes) [inline django/http/response.py]: Ahh right! I misread this as `upper()` somehow.

File: django/contrib/messages/storage/cookie.py [hunk 9]
```diff
@@ -86,6 +86,7 @@ def _update_cookie(self, encoded_data, response):
                 domain=settings.SESSION_COOKIE_DOMAIN,
                 secure=settings.SESSION_COOKIE_SECURE or None,
                 httponly=settings.SESSION_COOKIE_HTTPONLY or None,
+                samesite=settings.SESSION_COOKIE_SAMESITE or None,
```
Comment (jdufresne) [inline django/contrib/messages/storage/cookie.py]: According to the code comment, `SESSION_COOKIE_SAMESITE` can be `'lax'`, `'strict'` or `None`. When `SESSION_COOKIE_SAMESITE` is set to `None`, does this evaluate to `None or None`? If so, I don't think we need the `or None`, right? (Same in `django/contrib/sessions/middleware.py`)

Comment (kravietz) [inline django/contrib/messages/storage/cookie.py]: Isn't this `or None` intended for the case when the `settings.SOMETHING` isn't defined at all?

File: django/http/response.py (hunk 4 above)
Comment (jdufresne) [inline django/http/response.py]: Why not just expect the setting to be the correct capitalization, `'Lax'` or `'Strict'`? Doing `capitalize()` feels like unnecessary post-processing to me. Is it a matter of setting aesthetics?

**Comment by alex** [comment]
This probably deserves a mention in the release notes, I don't remember how we maintain those these days.

File: django/http/response.py (hunk 4 above)
Comment (kravietz) [inline django/http/response.py]: Fair point, corrected this.

**Comment by kravietz** [comment]
Release note added to 1.11.1.

**Comment by charettes** [comment]
@kravietz the code looks good but until _SameSite_ support lands in the stdlib I'm afraid we'll have to adapt [our vendored version of cookies](https://github.com/django/django/blob/master/django/http/cookie.py).

**Comment by kravietz** [comment]
@charettes Any idea what would be the best approach? The factory Morsel won't serialize anything that is not in the _reserved dict. I have started patching it a bit like it was done with the Django's SimpleCookie but with Morsel it would be much more ugly, essentially rewriting half of the BaseCookie class...

**Comment by charettes** [comment]
@kravietz I see, I really can't think of any other solution unfortunately.

**Comment by kravietz** [comment]
@charettes Actually, the patching didn't seem to be that hard. It seems to be working now. Also added a test specifically for the patched SimpleCookie and Morsel version.

**Comment by script3r** [comment]
@kravietz looking for help with this patch series? this helps reduce Spectre's attack surface. I'd love to help get this merged upstream.

**Comment by alex** [comment]
@script3r This hasn't been updated for a while, if you're interested in working on it I think you're good to take this and run with it.

**Comment by lane-c-wagner** [comment]
Is there anyway to set the flag currently in django 1.11? Some kind of workaround?

**Comment by alex** [comment]
Rebased and submitted as #9860

**Pull request closed without merging**